src/Manifest/ManifestManager/Options/OutTable/ManifestOptions.php

```diff
@@ -0,0 +1,357 @@
+declare(strict_types=1);
+
+namespace Keboola\Component\Manifest\ManifestManager\Options\OutTable;
+
+use Keboola\Component\Manifest\ManifestManager\Options\OptionsValidationException;
+use Keboola\Component\Manifest\ManifestManager\Options\OutTable\Serializer\LegacyManifestConverter;
+use Keboola\Component\Manifest\ManifestManager\Options\OutTable\Serializer\LegacyManifestNormalizer;
+use Keboola\Component\Manifest\ManifestManager\Options\OutTable\Serializer\NewNativeTypesManifestConverter;
+use Symfony\Component\Serializer\Encoder\JsonEncoder;
+use Symfony\Component\Serializer\Normalizer\AbstractObjectNormalizer;
+use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
+use Symfony\Component\Serializer\Serializer;
+use function array_keys;
+use function gettype;
+use function is_array;
+
+class ManifestOptions
+{
+    private string $destination;
+
+    /** @var string[] */
+    private array $primaryKeyColumns;
+
+    /** @var string[] */
+    private ?array $columns = null;
+
+    private bool $incremental;
+
+    /** @var mixed[][] */
+    private ?array $metadata = null;
+
+    /** @var mixed $columnMetadata */
+    private $columnMetadata = null;
+
+    private string $delimiter;
+
+    private string $enclosure;
+
+    /** @var ManifestOptionsSchema[]  */
+    private ?array $schema = null;
+
+    private ?string $manifestType = null;
+
+    private ?bool $hasHeader = null;
+
+    private ?string $description = null;
+
+    private ?array $tableMetadata = null;
+
+    private static function getLegacySerializer(): Serializer
+    {
+        $normalizer = new ObjectNormalizer(
+            null,
+            new LegacyManifestConverter(),
+            null,
+            null,
+            null,
+            null,
+            [AbstractObjectNormalizer::SKIP_NULL_VALUES => true],
+        );
+        $encoders = [new JsonEncoder()];
+
+        return new Serializer([new LegacyManifestNormalizer($normalizer), $normalizer], $encoders);
+    }
+
+    private static function getNewNativeTypesSerializer(): Serializer
+    {
+        $nameConverter = new NewNativeTypesManifestConverter();
+        $normalizer = new ObjectNormalizer(null, $nameConverter);
+        $encoders = [new JsonEncoder()];
+
+        return new Serializer([$normalizer], $encoders);
+    }
+
+    public function toArray(bool $legacy = true): array
+    {
+        $serializer = $legacy ? self::getLegacySerializer() : self::getNewNativeTypesSerializer();
+
+        return (array) $serializer->normalize($this, null, [AbstractObjectNormalizer::SKIP_NULL_VALUES => true]);
+    }
+
+    public static function fromArray(array $data, bool $legacy = true): ManifestOptions
+    {
+        if ($legacy) {
+            $serializer = self::getLegacySerializer();
+        } else {
+            $serializer = self::getNewNativeTypesSerializer();
+        }
+
+        /** @var ManifestOptions $manifestOptions */
+        $manifestOptions = $serializer->denormalize($data, self::class);
```

Comment:
No podle mě sis zadřel nedomrlostí téhle metody/classy. Protože ManifestOptions má columns a schema tak nikdy nevíš co z toho vzít a kdy použít. 
Což je zbytečná komplikace podle mě. Novej manifest je superset starýho manifestu, takže ManifestOptions by měla odpovídat novýmu (tj. nemít columns) - a tady ve fromArray uděláš potřebnou konverzi - z legacy columns uděláš schema (což vždycky jde).

Při uložení pak buď uložíš schéma nebo to zase zkonvertuješ obráceně, podle toho, co se má udělat.